### PR TITLE
More scribble documentation, bug fixes, and deployment system improvements

### PR DESCRIPTION
**Pull request by col6y**
@amackworth - would you take a look?

### CODE REVIEW

File: CommonChickenRuntimeEngine/src/ccre/concurrency/ReporterThread.java [hunk 1]
```diff
@@ -69,7 +69,7 @@ public final void run() throws IllegalStateException {
         } catch (OutOfMemoryError oom) {
             System.err.println("OutOfMemory");
             Logger.severe("OutOfMemory");
-            throw oom; // The out-of-memory error will crash the system, at least in the Squawk VM.
+            Logger.severe("OutOfMemory", oom); // in case we can
```
Comment (Ja-ake) [inline CommonChickenRuntimeEngine/src/ccre/concurrency/ReporterThread.java]: Well, I'm not amackworth, but I'll take a look at all of this.

Is there a reason why you're printing "OutOfMemory" twice to the Logger's severe channel? The OOMError is appended to the second print statement, but the message "OutOfMemory" still printed.

File: DeploymentEngine/src/ccre/deployment/DepRoboRIO.java [hunk 2]
```diff
@@ -261,8 +264,8 @@ public int getRIOImage() throws IOException {
     }
 
     public static Artifact build(File source, Class<? extends FRCApplication> main) throws IOException {
-        Artifact newcode = DepJava.build(source, DepRoboRIO.getJarFile());
-        return DepJar.combine(DepRoboRIO.manifest(main), JarBuilder.DELETE, newcode, DepRoboRIO.getJar());
+        Artifact newcode = DepJava.build(source, DepRoboRIO.getJarFile(true));
+        return DepJar.combine(DepRoboRIO.manifest(main), JarBuilder.DELETE, newcode, DepRoboRIO.getJar(false));
```
Comment (Ja-ake) [inline DeploymentEngine/src/ccre/deployment/DepRoboRIO.java]: Why true in the line above and why false in this one? Quick comment will help people quickly see what this code is doing

File: Emulator/src/ccre/frc/DeviceListMain.java [hunk 3]
```diff
@@ -43,24 +42,13 @@
...
-    public static void main(String[] args) throws IOException, ClassNotFoundException, InstantiationException, NoSuchMethodException, IllegalAccessException, InvocationTargetException, InterruptedException {
-        if (args.length != 1) {
-            System.err.println("Expected arguments: <FRC-Jar>");
-            System.exit(-1);
-            return;
-        }
-        File jarFile = new File(args[0]);
+    public static void startEmulator(File jarFile, File storageDir) throws Exception {
         JarFile frcJar = new JarFile(jarFile);
+        Storage.setBaseDir(storageDir);
```
Comment (Ja-ake) [inline Emulator/src/ccre/frc/DeviceListMain.java]: Is this sanitized and checked for validity?

File: Emulator/src/ccre/frc/DeviceListMain.java [hunk 4]
```diff
@@ -43,24 +42,13 @@
...
      */
-    public static void main(String[] args) throws IOException, ClassNotFoundException, InstantiationException, NoSuchMethodException, IllegalAccessException, InvocationTargetException, InterruptedException {
-        if (args.length != 1) {
-            System.err.println("Expected arguments: <FRC-Jar>");
-            System.exit(-1);
-            return;
-        }
-        File jarFile = new File(args[0]);
+    public static void startEmulator(File jarFile, File storageDir) throws Exception {
         JarFile frcJar = new JarFile(jarFile);
```
Comment (Ja-ake) [inline Emulator/src/ccre/frc/DeviceListMain.java]: Is this sanitized and checked for validity?

File: CommonChickenRuntimeEngine/src/ccre/concurrency/ReporterThread.java (hunk 1 above)
Comment (col6y) [inline CommonChickenRuntimeEngine/src/ccre/concurrency/ReporterThread.java]: In the case of an OOM, it might not be possible to print out an exception, so we try to print something  shorter first.

File: docsrc/index.scrbl [hunk 5]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+and later:
+
+@codeblock|{
+  EventOutput pet_fluffy = new CatPettingEventOutput(fluffy);
+    }|
+
+But then the class definition is far away from the actual use, and your code gets clogged up with all of your classes. Clearly, there has to be an easier way:
+
+@codeblock|{
+     EventOutput pet_fluffy = new EventOutput() {
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: Make this example class consistent with the CatPettingEventOutput class.

File: DeploymentEngine/src/ccre/deployment/DepRoboRIO.java (hunk 2 above)
Comment (col6y) [inline DeploymentEngine/src/ccre/deployment/DepRoboRIO.java]: I'll add some constants for true and false to make it more clear.

File: docsrc/index.scrbl [hunk 6]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+     };
+     BooleanOutput fluffy_cage_light = (light_on) -> {
+       // turn on or off the light in fluffy's cage
+     };
+     FloatOutput fluffy_heater_temp = (temperature) -> {
+       // change the thermostat on fluffy's cage
+     };
+     }|
+
+(Also, if you only have a single statement in the lambda, you can omit the @code|{{}}| and the semicolon around the statement.)
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: Show an example of this.

File: docsrc/index.scrbl [hunk 7]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+It also talks to the Field Management System (FMS) if you're playing in a real competition.
+
+The software looks like this:
+
+@image["driver-station.png"]
+
+The laptop might looks something like this: (depending on your team)
+
+@image["driver-station-laptop.jpg"]
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: I can't comment on the image directly, but I find the computer screen to be really bright and the computer to be really dark. You could either retake a better image using a shorter exposure time in a brighter area or you could just remove this, it isn't really needed.

File: docsrc/index.scrbl [hunk 8]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+@margin-note{I'm only showing the left half of the DS - the right half is relatively unimportant.}
+
+Let's go over the tabs available on the DS:
+
+@image["driver-station-annotated.png"]
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: The annotation text is hard to see, consider changing the color to something brighter.

File: docsrc/index.scrbl [hunk 9]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+@subsubsection{Keyboard shortcuts}
+
+There are a few important keyboard shortcuts:
+
+@itemlist[@item{The spacebar is the @italic{emergency-stop} button. Once you press it, the robot stops running until you reboot it physically.}
+          @item{The enter key is the disable key. It is the same as pressing the disable button on the first tab of the driver station.}
+          @item{If you press the keysequence @code|{[]\ }|, the robot will enable. This is the same as pressing the enable button on the first tab of the driver station.}]
+
+Always keep your hand near the disable key when enabling the robot.
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: And let people around you know that you are enabling.

File: docsrc/index.scrbl [hunk 10]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+This is ALSO a Joystick:
+
+@image["joystick-flight.jpg"]
+
+This? Another Joystick. Not two Joysticks - just one.
+
+@image["joystick-xbox.jpg"]
+
+Each Joystick has some number of @italic{axes} - each axis measures a value from -1.0 to 1.0. It also has some number of @italic{buttons} - each button can be either pressed or unpressed.
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: "Buttons can be pressed or unpressed"[citation needed]

Comment (Ja-ake) [inline docsrc/index.scrbl]: Also the term is "released"

File: Emulator/src/ccre/frc/DeviceListMain.java (hunk 3 above)
Comment (col6y) [inline Emulator/src/ccre/frc/DeviceListMain.java]: Hmm. No, but that should be the job of Storage.setBaseDir.

File: docsrc/index.scrbl [hunk 11]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+Standard axes are the position of a Joystick on the forward-backward (Y) axis and the left-right (X) axis.
+
+For example, a trigger on a Joystick is a button, and an altitude control wheel (as on the base of the first Joystick) is an axis.
+
+The X axis on a Joystick is usually axis #1, and the Y axis is usually axis #2.
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: X axis is axis 1 or 0? This could confuse.

File: Emulator/src/ccre/frc/DeviceListMain.java (hunk 4 above)
Comment (col6y) [inline Emulator/src/ccre/frc/DeviceListMain.java]: JarFile checks for non-nullity and existence, but you are correct in that it should be checked here as well.

File: docsrc/index.scrbl (hunk 5 above)
Comment (col6y) [inline docsrc/index.scrbl]: What exactly do you think is different?

File: docsrc/index.scrbl [hunk 12]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+If you work with something with multiple XY sticks, this may vary. On an xbox controller, for example, this is the left stick, but for the right stick, the X axis is #5 and the Y axis is #6. (And the triggers are #3 (left) and #4 (right.))
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: "this is the left stick" should be reworded to be more clear what "this" is.

File: docsrc/index.scrbl [hunk 13]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+It contains a set of ports for interfacing with PWM-controlled devices, CAN bus devices, I@superscript{2}C devices, SPI devices, RS232 (serial) devices, miscellaneous digital I/O devices, relays, analog inputs, USB devices, and networked devices over Ethernet.
+
+You aren't allowed to control anything on your robot (except for nonfunctional components like LEDs) with any other controller than the roboRIO, so teams have to use it as their main controller.
+
+See below in this document for some of the devices that attach to the roboRIO.
+
+@subsubsection{Hardware access}
+
+The CCRE provides interfaces to the underlying hardware via WPILib's JNI layer, which attaches to the WPILib Hardware Abstraction Layer (in C++), which attaches to the NI ChipObject propriatary library, which attaches to the NiFPGA interface to the FPGA (field-programmable gate array) device that manages the communication between the higher-level code and the I/O ports.
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: What's WPILib? What's JNI?

Comment (Ja-ake) [inline docsrc/index.scrbl]: What is a "hardware abstraction layer"

File: docsrc/index.scrbl (hunk 7 above)
Comment (col6y) [inline docsrc/index.scrbl]: It's not that bad on any of my computers. It might just be your monitor. I don't see any harm in leaving it in. If you'd like to retake a photo, feel free, but I find the current one to be good enough.

File: docsrc/index.scrbl [hunk 14]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+@vl-append[10
+ (ht-append 10
+            (vl-append 10
+                       (ht-append 10 (vl-append 10
+                                                (ht-append 10 (cellw "CCRE Hardware I/O" 150) (cellw "WPILibJ" 150))
+                                                (cellw "WPILib JNI Layer" 310))
+                                  (cellw "WPILibC++" 150 90)
+                                  (cellw "pyfrc" 150 90))
+                       (cellw "Hardware Abstraction Layer (HAL)" 630 40)
+                       (cellw "Ni ChipObject .so" 630 40)
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: ".so" doesn't mean anything to a non-linux-friendly user.

File: docsrc/index.scrbl (hunk 8 above)
Comment (col6y) [inline docsrc/index.scrbl]: Hmm. Good point. I may have to redo this, as I think I might have lost the source .xcf.

File: docsrc/index.scrbl (hunk 9 above)
Comment (col6y) [inline docsrc/index.scrbl]: Don't I say something about that in the Safety section?

File: docsrc/index.scrbl (hunk 10 above)
Comment (col6y) [inline docsrc/index.scrbl]: Both are valid terminology. You are correct in that it might be better to say "released". However, noting that they are binary, while obvious to us, does help prevent confusion. It also clarifies that it's not just an event for button press and nothing for button release.

File: docsrc/index.scrbl (hunk 6 above)
Comment (col6y) [inline docsrc/index.scrbl]: Will do.

File: docsrc/index.scrbl [hunk 15]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+@image["deploy.png"]
+
+But how does this work behind the scenes?
+
+First, this goes through the DeploymentEngine, which dispatches to the @code{deploy()} DepTask in the default Deployment class:
+
+@codeblock|{
+    @DepTask
+    public static void deploy() throws Exception {
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: Add (really really good) clarifying comments to this code

File: docsrc/index.scrbl (hunk 11 above)
Comment (col6y) [inline docsrc/index.scrbl]: #1.

File: docsrc/index.scrbl (hunk 12 above)
Comment (col6y) [inline docsrc/index.scrbl]: Will do.

File: docsrc/index.scrbl (hunk 13 above)
Comment (col6y) [inline docsrc/index.scrbl]: Perhaps I should move this section into the 'maintainer' section, as it's mostly trivia.

File: docsrc/index.scrbl (hunk 14 above)
Comment (col6y) [inline docsrc/index.scrbl]: Do you have a better abbreviation?

File: docsrc/index.scrbl (hunk 15 above)
Comment (col6y) [inline docsrc/index.scrbl]: Perhaps I should also move some of this to the maintainer section. And, yes, add clarifying comments.

Comment (col6y) [inline docsrc/index.scrbl]: Actually, on second thought, I should move it to the detailed CCRE section, not to the maintainer section.

File: docsrc/index.scrbl [hunk 16]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+On Linux or Mac OS X, you can do this from the command line with pre-installed tools:
+
+@margin-note{On Windows, you can use PuTTY.}
+@codeblock|{
+ $ ssh admin@roboRIO-1540.local
+ Password:
+ $ do stuff on the robot
+}|
+
+The default password is the blank password, and the default username (instead of 'root') is 'admin'.
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: You can get rid of the "instead of root" I think

Comment (col6y) [inline docsrc/index.scrbl]: It's relevant to anyone who expects the superuser/uid 0 on a Linux system to be 'root'. However, I could clarify why I'm mentioning this.

File: docsrc/index.scrbl [hunk 17]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+@image["motor-talon-sr.jpg"]
+
+A Talon SRX (also controllable over CAN):
+
+@image["motor-talon-srx.jpg"]
+
+For a discussion of how PWM works, see @hyperlink["https://en.wikipedia.org/wiki/Pulse-width_modulation"]{the Wikipedia article}. The important attributes for us are:
+
+@itemlist[@item{PWM as a protocol is unidirectional. You can't get any information from a PWM device.}
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: I would rephrase as "a PWM device can't send you any data."

Comment (col6y) [inline docsrc/index.scrbl]: Good idea.

**Comment by Ja-ake** [comment]
I'll finish up later.

**Comment by col6y** [comment]
Sounds good.

File: docsrc/index.scrbl (hunk 5 above)
Comment (col6y) [inline docsrc/index.scrbl]: Talked in person. Concerns mitigated.

File: docsrc/index.scrbl (hunk 8 above)
Comment (col6y) [inline docsrc/index.scrbl]: I have split off #75.

File: docsrc/index.scrbl (hunk 9 above)
Comment (col6y) [inline docsrc/index.scrbl]: I do.

File: docsrc/index.scrbl [hunk 18]
```diff
@@ -413,29 +416,617 @@ See @secref["remixing"] for more info.
...
+
+For example, the IP address @code{192.168.1.1} is a valid local IP address. In the FRC control system, IP addresses are assigned based on team number, in the form @code{10.TE.AM.0/24} (which means, for example, @code{10.15.40.0} through @code{10.15.40.255} for team 1540. For shorter team numbers, it might be @code{10.2.54.XX} or @code{10.0.11.XX}.)
+
+Within that range, certain addresses are reserved: (for this example, we use team 1540 addresses.)
+
+@codeblock|{
+10.15.40.0 is reserved (addresses ending in 0 usually are) - don't use it.
+10.15.40.1 is the IP address of the robot's wireless access point. This is usually a D-Link DAP-1522 Rev B wireless radio.
+10.15.40.2 used to be reserved for the cRIO. It can be used by the roboRIO if you wish.
+10.15.40.2 through 10.15.40.19: unused.
```
Comment (Ja-ake) [inline docsrc/index.scrbl]: I think you mean 10.15.40.3

**Comment by col6y** [comment]
@Ja-ake or @amackworth: ready for re-review.

File: CommonChickenRuntimeEngine/src/ccre/concurrency/ReporterThread.java (hunk 1 above)
Comment (Ja-ake) [inline CommonChickenRuntimeEngine/src/ccre/concurrency/ReporterThread.java]: Clever.

File: docsrc/index.scrbl (hunk 7 above)
Comment (Ja-ake) [inline docsrc/index.scrbl]: It'll suffice for now at least, it probably just is my monitor.

File: docsrc/index.scrbl (hunk 14 above)
Comment (Ja-ake) [inline docsrc/index.scrbl]: "lib", "library"? I don't know, .so is fine since anyone who doesn't know what it is probably shouldn't be concerned with Ni ChipObject.

File: docsrc/index.scrbl (hunk 15 above)
Comment (Ja-ake) [inline docsrc/index.scrbl]: Sounds good.
